Add backend exception handling and Un_cps design docs (#191)
ocaml-flambda/flambda-backend: middle_end/flambda2/docs/to_cmm.md

these tasks are implemented will be discussed in the following section).

The main challenges of the translation are:
- translate some abstract concepts into concrete ones, for such things as
  tagging, boxing, "big" and "small" values (aka Int64 and Int32 depending on
  architecture), etc...
- translating the continuations of flambda2 into the control flow structures of
  the cmm language
- substituting some let-bindings from the flambda2, so as to generate (at least for
  arithmetic) more nested expressions which are better optimized by the
  construction function for cmm
- assigning unique offsets to all elements in set of closures (closures and env
  vars), while taking into account offsets assigned by other compilation units
- eliminate unused env vars for sets of closures, sometimes resulting in a set
  of closure that can be statically allocated
- try and avoid using a global mutable state, but still try and re-use most of
  the existing code as possible

### Low-level implementation of high-level concepts

The points here (tagging/unboxing, and big/small values) are actually
tasks that are also performed by cmmgen (because they are properties of the

> diffrecnes between representtion between middle-end and backend).

Author: fixed

#### Tagging and boxing

As it is the turning point between the middle-end and the backend of the
compiler, some operations/concepts that are abstract in earlier passes need
to be made far more explicit.

First is to introduce all the concrete tagging/untagging and boxing/unboxing.
Flambda2 already distinguishes tagged and untagged integers as well as boxing
(in order to be able to reason about those and eliminate them when possible),
however those are still abstract at the flambda2 level, whereas at the Cmm level
and lower these operations are explicit, e.g.:
- cmm only operates on machine integers, so tagging explicitly has to be
  translated in terms of arithmetic (left shift and addition for tagging for
  instance)
- allocations still benefit from some builtin handling, but block headers
  are integers from cmm onwards, so concrete integer values for the headers
  must be computed (whereas before, it was only the tag info)
- field projection must be translated as a memory address computation
  (pointer of the value + offset of the field, which can depend on the integer
  size for the targeted architecture), and then a memory read

> - similarly, field assignements must be expressed as memory address

Reviewer: ```suggestion
- similarly, field assignments must be expressed as memory address
```

restricted to the current module, there should not be any instance where
the offset computation can fail.

Currently, `un_cps` implements a greedy algorithm for filling out sets of
closures and assign offsets. It iterates over closures, assigning the lowest
offset that is free in all sets of closures where it appears, and then does the
same for all env vars. While not optimal as it may leave gaps, this should
result in the regular solution for when there is no sharing between sets of
closures (thus preserving the behaviour of Closure), behave well in cases
with reasonable sharing (mainly when sharing occurs between sets of closures
defined by the same module), and provide a best effort for exotic cases with
sharing across modules, all the while having reasonable complexity.

[1]: Actually, Flambda1 had a fallback that allowed dynamic projection, where
the projection would, at runtime, look at the set of closure in order to
determine which field to read, but we'd very much like not to have that in
flambda2.

## Concrete details

### Organization of the code and main control flow of the translation

`un_cps` translate flambda2 expressions using two traversals of an flambda2
body.

> The first traverssal iterates over all sets of closures defined by the body.

Reviewer: ```suggestion
The first traversal iterates over all sets of closures defined by the body.
```

This allows to assign offsets to all members of sets of closures (taking into
account constraints due to a potential sharing of closures of env vars beetween
sets of closures), using the code in `un_cps_closure.ml`.

After that, `un_cps` operates the second, and main traversal of the
expression, accumulating information (about continuations, substitutions, etc)
in an environment (defined in `un_cps_env.ml`), and then returning two
values: the resulting cmm expression, and a static result that contains all
the constants that should be pre-allocated (in comparison, `cmmgen` uses a global
reference to store that information during translation). This static result
(defined in `un_cps_result.ml`) will be grown by the translation, by either
`let_symbol` bindings or sets of closures bindings. Translations of flambda2
expression that can result in pre-allocation of static constants go through
the `un_cps_static.ml` file.

### Tagging and unboxing

In the code, this is one of the rather easy points. Flambda2 has a notion

> of kinds, which offser enough information about the memory representation

Reviewer: ```suggestion
of kinds, which offers enough information about the memory representation
```

of values, e.g. boxed integers are identified with their size, tagged integers
are differentiated from untagged integers, ... . Together with information
about what the cmm primitives expect (i.e. which primitive expect or return
a tagged integer, and which deal with untagged integers), this offers a
rather comprehensive way for `un_cps` to know when or where to tag/untag
things. See the following for more details:
- `middle_end/flambda/terms/flambda_primitive.ml` for the data about kinds
  consumed/returned by flambda primitives (particularly the
  `arg_kind_of_*_primitive` and `result_kind_of_*_primitive`)
- [the cmm doc](cmm.md) for info about the primitives in cmm
  (this is still a TODO)

> Unboxing is also straightforward: as mentioneed earlier, flambda2 already

Reviewer: ```suggestion
Unboxing is also straightforward: as mentioned earlier, flambda2 already
```

takes care of identifying and applying all the unboxing we want, so `un_cps`
only introduces the boxing and unboxing as needed, using the same information
as for tagging/untagging.

For the concrete implementation, `un_cps` mostly re-uses code from
`cmm_helpers`.

### "Big" and "small" values

We'll describe here the way "big" and "small" values are translated, since
the rest is relatively simple and can be understood directly from the code.

#### Int64 on 32-bit archs

Since in most cases, these are handled by C functions, the C convention is used

> throughtou all the code: on a 32-bit arch, an Int64 is split into its lower

Reviewer: ```suggestion
throughout all the code: on a 32-bit arch, an Int64 is split into its lower
```

> bits and its higher bits, and then carrried on two registers.  Handling these

Reviewer: ```suggestion
bits and its higher bits, and then carried on two registers.  Handling these
```

range (division, modulo, etc...), but these were already implemented in
`cmm_helpers`.

### Continuations

As stated before, continuations are translated in one of two ways:
- non-recursive continuations that are used exactly once are inlined
- all other continuations are translated using static jump/catches

Inlined continuations are first stored untranslated (i.e. the handler of
the continuation is stored as a flambda2 expression), so that when it is
inlined, the translation of the expression can make use of substitutions.

Non-inline continuations (or jumps), are pretty straight-forward since cmm
has a notion of static jump and catch. The only thing to pay attention to
is to check the potential trap actions that can be associated to a continuation
call. But again, these are relatively simple to translate since cmm now has
a trap mechanism to mirror that of flambda2.

### Let-binding substitution

This is the most complex part of `un_cps` and also the part most sensitive
to changes (even innocuous changes in `un_cps` might break becasue of this).
The goal is to substitute let-bound variable's bodies during translation, because

> it allows for more optimizations. There are two technical challengs

Reviewer: ```suggestion
it allows for more optimizations. There are two technical challenges
```

with this:
- keep the correct order of evaluation (with respect to effects and co-effects)
- yet still allow for some re-ordering of evaluation, such as pure computations
  across other computations (as long as it doesn't cross into a recursive

>   continuations, in which case the expression might be valuated more than once).

Reviewer: ```suggestion
  continuations, in which case the expression might be evaluated more than once).
```

Most of the code to deal with this is in the definition of the translation
environment in `un_cps_env`: the environment accumulates all let-bindings
(whether that can be substituted or not), since we have to be able to
substitute through a non-substitutable binding. A nice consequence of that is
that translating a let-binding can then be tail-rec, which is very useful.

Let-bindings are stored in the environment in a few structures:
- pure bindings are stored in a dedicated map
- all other bindings are stored in "stages". Intuitively a stage is a set of bindings
  that can all be commuted with each other, while keeping the same semantics. This
  means that a stage is either: a set of let-bindings which only have co-effects,
  or a single effectful binding.
These two structures are used to store enough information to perform substitution
if needed. Additionally, another map is used to relate all flambda2 variables to

> a corresponding cmm variables. This mapping is used when `un_cps` decides not to

Reviewer: ```suggestion
corresponding cmm variables. This mapping is used when `un_cps` decides not to
```

the variable is marked as substitutable, else the generic mapping from flambda2
variables to cmm variables is used. On variables bound to a non-pure body, the
last and current stage is looked up, if the variable is in it and can be
substituted (i.e. has been marked as substitutable when inserting it in the
env[2]) the substitued body is returned, else the generic mapping is used. Is
the variable has been substituted, it is then removed from the current stage
and if this results in the current stage being empty, the previous stage is
restored as the current stage.

[2]: this is done by the caller of `Env.bind_variable`, on variables that are used
exactly once.

This process allows to correctly substitute let-bound variable's bodies when
adequate, but for variables not substituted, the let-binding information is
"only" in the env: when translating the let-binding, all that is done is add the
let-binding info in the env, thus if the variable's body is not substituted, it
simply stays in the env. Hence, the env needs to be "flushed" of its accumulated
let-bindings at some precise points to ensure that all let-bindings are correctly

> translated and present in the generated cmm code. This presents two challenge:

Reviewer: ```suggestion
translated and present in the generated cmm code. This presents two challenges:
```

- correctly identify all points where the env needs to be flushed (else variables
  may be unbound)
- the env must be treated as a somewhat linear type, more precisely an env can
  only be duplicated right after a flush and before any new variables are bound
  in it (else this may result in duplicated bindings being generated).

The points at which the env should be flushed can basically be understood as any
operations that cannot be expressed as a let-binding, so any control flow,
including non-inlined continuations, but excluding function calls (since these can
be seen as effectful let-bindings).

### Closure offsets

Closure offset computation is done in two steps. First is to accumulate all

> information about sets of closures and members of these in ia mutable

Reviewer: ```suggestion
information about sets of closures and members of these in a mutable
```

> structure: by iterating over all sets of closure sin the flamdba2 program body,

Reviewer: ```suggestion
structure: by iterating over all sets of closures in the flamdba2 program body,
```

> we build a record for each set of closures. These records contains a list of

Reviewer: ```suggestion
we build a record for each set of closures. These records contain a list of
```

closures and of env vars that belong to the set of closures; we will refer to
them as slots. A slot is a member of a set of closures, currently a slot is

> occupied by either a closure or an env var.  Additionally, each slot record all

Reviewer: ```suggestion
occupied by either a closure or an env var.  Additionally, each slot records all
```
